uvtools/dspec.py

```diff
@@ -73,36 +86,87 @@ def high_pass_fourier_filter(data, wgts, filter_size, real_delta, tol=1e-9, wind
+        assert len(real_delta) == 2, "len(real_delta) must equal 2 for 2D clean"
+    else:
+        assert isinstance(real_delta, (int, np.integer, float, np.float)), "if not twoD, real_delta must be a float"
+
+    # 1D clean
+    if not twoD:
+        # setup _d and _w arrays
+        nchan = data.shape[-1]
+        _window = np.empty(nchan)
+        _window[:] = aipy.dsp.gen_window(nchan, window=window, **win_kwargs)
+        if dndim == 2:
+            _window = _window[None, :]
+        _d = np.fft.ifft(data * wgts * _window, axis=-1)
+        _w = np.fft.ifft(wgts * _window, axis=-1)
+
+        # calculate area array
+        area = np.ones(nchan, dtype=np.int)
+        uthresh, lthresh = calc_width(filter_size, real_delta, nchan)
+        area[uthresh:lthresh] = 0
+
+        # run clean
+        if dndim == 1:
+            # For 1D data array run once
+            _d_cl, info = aipy.deconv.clean(_d, _w, area=area, tol=tol, stop_if_div=False, maxiter=maxiter, gain=gain)
+            d_mdl = np.fft.fft(_d_cl)
+            del info['res']
+        elif data.ndim == 2:
+            # For 2D data array, iterate
+            info = []
+            d_mdl = np.empty_like(data)
+            for i in range(data.shape[0]):
+                if _w[i, 0] < skip_wgt:
+                    d_mdl[i] = 0 # skip highly flagged (slow) integrations
+                    info.append({'skipped': True})
+                else:
+                    _d_cl, info_here = aipy.deconv.clean(_d[i], _w[i], area=area, tol=tol, stop_if_div=False, maxiter=maxiter, gain=gain)
+                    d_mdl[i] = np.fft.fft(_d_cl)
+                    del info_here['res']
+                    info.append(info_here)
+
+    # 2D clean on 2D data
+    else:
+        # setup _d and _w arrays
+        _window = np.empty(data.shape[0])
+        _window[:] = aipy.dsp.gen_window(data.shape[0], window=window, **win_kwargs)
+        _window = _window[:, None]
+        _d = np.fft.ifft(data * wgts * _window, axis=0)
+        _w = np.fft.ifft(wgts * _window, axis=0)
+        _window = np.empty(data.shape[1])
+        _window[:] = aipy.dsp.gen_window(data.shape[1], window=window, **win_kwargs)
+        _window = _window[None, :]
+        _d = np.fft.ifft(_d * _window, axis=1)
+        _w = np.fft.ifft(_w * _window, axis=1)
+
+        # calculate area array
+        a1 = np.ones(data.shape[0], dtype=np.int)
+        uthresh, lthresh = calc_width(filter_size[0], real_delta[0], data.shape[0])
+        a1[uthresh:lthresh] = 0
+        a2 = np.ones(data.shape[1], dtype=np.int)
+        uthresh, lthresh = calc_width(filter_size[1], real_delta[1], data.shape[1])
+        a2[uthresh:lthresh] = 0
+        area = np.outer(a1, a2)
+
+        # run clean
         _d_cl, info = aipy.deconv.clean(_d, _w, area=area, tol=tol, stop_if_div=False, maxiter=maxiter, gain=gain)
-        d_mdl = np.fft.fft(_d_cl)
+        d_mdl = np.fft.fft(np.fft.fft(_d_cl, axis=0), axis=1)
```

Comment:
well, the function still returns the residual, just in frequency space. so one can still reconstruct this themselves if they want, I think..

Reply:
I think it might be more accurate, since this is including the part of the residual that's inside the filtered region. @AaronParsons convinced me that it's slightly better to include it.

Reply:
Also, it's not easily reconstructed, since the returned d_res isn't the same thing.